crates/dbs-arch/src/aarch64/pmu.rs

```diff
@@ -0,0 +1,172 @@
+    SetPmuInit(#[source] KvmError),
+}
+
+type Result<T> = std::result::Result<T, PmuError>;
+
+/// Tests whether a cpu supports KVM_ARM_VCPU_PMU_V3_IRQ attribute.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn has_pmu_irq(vcpu: &VcpuFd) -> Result<()> {
+    let irq = (VIRTUAL_PMU_IRQ + PPI_BASE) as u64;
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_IRQ),
+        addr: &irq as *const u64 as u64,
+        flags: 0,
+    };
+    vcpu.has_device_attr(&attribute)
+        .map_err(PmuError::HasPmuIrq)
+}
+
+/// Tests whether a cpu supports KVM_ARM_VCPU_PMU_V3_INIT attribute.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn has_pmu_init(vcpu: &VcpuFd) -> Result<()> {
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_INIT),
+        addr: 0,
+        flags: 0,
+    };
+    vcpu.has_device_attr(&attribute)
+        .map_err(PmuError::HasPmuInit)
+}
+
+/// Set KVM_ARM_VCPU_PMU_V3_IRQ for a specific vcpu.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn set_pmu_irq(vcpu: &VcpuFd) -> Result<()> {
+    let irq = (VIRTUAL_PMU_IRQ + PPI_BASE) as u64;
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_IRQ),
+        addr: &irq as *const u64 as u64,
+        flags: 0,
+    };
+    vcpu.set_device_attr(&attribute)
+        .map_err(PmuError::SetPmuIrq)
+}
+
+/// Set KVM_ARM_VCPU_PMU_V3_INIT for a specific vcpu.
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn set_pmu_init(vcpu: &VcpuFd) -> Result<()> {
+    let attribute = kvm_device_attr {
+        group: KVM_ARM_VCPU_PMU_V3_CTRL,
+        attr: u64::from(KVM_ARM_VCPU_PMU_V3_INIT),
+        addr: 0,
+        flags: 0,
+    };
+    vcpu.set_device_attr(&attribute)
+        .map_err(PmuError::SetPmuInit)
+}
+
+/// Check kvm pmu capability
+///
+/// # Arguments
+/// * `vm` - The VM file descriptor
+fn check_kvm_pmu_cap(_vm: &VmFd) -> Result<()> {
+    // TODO: check KVM_CAP_ARM_PMU_V3 capability before setting PMU
+    // Cap for KVM_CAP_ARM_PMU_V3 isn't supported in kvm-ioctls upstream, so
+    // leave a todo here for supporting this check in the future.
+    // Interface: vm.check_extension(kvm_ioctls::Cap)
+
+    Ok(())
+}
+
+/// Check pmu feature
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn check_pmu_feature(vcpu: &VcpuFd) -> Result<()> {
+    has_pmu_irq(vcpu)?;
+    has_pmu_init(vcpu)
+}
+
+/// Set pmu feature
+///
+/// # Arguments
+/// * `vcpu` - The VCPU file descriptor
+fn set_pmu_feature(vcpu: &VcpuFd) -> Result<()> {
+    set_pmu_irq(vcpu)?;
+    set_pmu_init(vcpu)
+}
+
+/// Initialize PMU in for vcpu
+///
+/// # Arguments
+/// * `vm` - The VM file descriptor
+/// * `vcpu` - The VCPU file descriptor
+pub fn initialize_pmu(vm: &VmFd, vcpu: &VcpuFd) -> Result<()> {
```

Comment:
Where should this function be triggered? I don't see in this PR

Reply:
It is used as a global interface to initialize pmu feature on aarch64 and will be triggered in dragonball vmm.